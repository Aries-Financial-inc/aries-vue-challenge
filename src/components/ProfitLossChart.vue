<template>
  <div :class="$style.Root">
    <LineChartJS :chartData="chartData" :chartOptions="chartOptions" />
  </div>
</template>

<script lang="ts">
import Vue from 'vue'

/* import models */
import { Option } from '@/models/Option'

/* import utils */
import { generateChartData } from '@/utils/OptionUtils'
import { formatPrice } from '@/utils/FormatUtils'
import { CHART_COLORS } from '@/utils/ChartUtils'

/* import components */
import { Line as LineChartJS } from 'vue-chartjs/legacy'

import {
  Chart as ChartJS,
  Title,
  Tooltip,
  Legend,
  LineElement,
  LinearScale,
  CategoryScale,
  PointElement,
  ChartData,
  ChartOptions
} from 'chart.js'

ChartJS.register(
  Title,
  Tooltip,
  Legend,
  LineElement,
  LinearScale,
  CategoryScale,
  PointElement
)

export default Vue.extend({
  components: {
    LineChartJS
  },

  props: {
    optionsData: {
      type: Array as () => Option[],
      required: true
    }
  },

  computed: {
    chartData(): ChartData {
      const { prices, profits } = generateChartData(this.optionsData)

      return {
        labels: [...prices],
        datasets: [
          {
            label: 'Profit / Loss',
            backgroundColor: CHART_COLORS.blue,
            borderColor: CHART_COLORS.blue,
            data: [...profits],
            segment: {
              borderColor: (ctx) => {
                const price = ctx.p1.parsed.y
                if (price <= 0) {
                  return CHART_COLORS.red
                }
              },

              backgroundColor: (ctx) => {
                const price = ctx.p0.parsed.y
                if (price <= 0) {
                  console.log('red', price)
                  return CHART_COLORS.red
                }
              }
            },
            spanGaps: true
          }
        ]
      }
    },

    chartOptions(): ChartOptions {
      return {
        responsive: true,
        maintainAspectRatio: false,
        plugins: {
          legend: {
            display: false
          }
        },
        scales: {
          x: {
            title: {
              display: true,
              text: 'Price of Underlying at Expiry'
            }
          },
          y: {
            title: {
              display: true,
              text: 'Profit / Loss'
            },
            ticks: {
              callback: (value) => {
                return formatPrice(value)
              }
            }
          }
        }
      }
    }
  }
})
</script>

<style module>
.Root {
  width: 100%;
  height: 400px;
}
</style>
