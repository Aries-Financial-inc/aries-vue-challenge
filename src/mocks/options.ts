export const sampleData = [
  {
    strike_price: 100,
    type: 'Call',
    bid: 10.05,
    ask: 12.04,
    position: 'Long',
    expiration_date: '2025-12-17T00:00:00Z'
  },
  {
    strike_price: 102.5,
    type: 'Call',
    bid: 12.1,
    ask: 14,
    position: 'Long',
    expiration_date: '2025-12-17T00:00:00Z'
  },
  {
    strike_price: 103,
    type: 'Put',
    bid: 14,
    ask: 15.5,
    position: 'Short',
    expiration_date: '2025-12-17T00:00:00Z'
  },
  {
    strike_price: 105,
    type: 'Put',
    bid: 16,
    ask: 18,
    position: 'Long',
    expiration_date: '2025-12-17T00:00:00Z'
  }
]
