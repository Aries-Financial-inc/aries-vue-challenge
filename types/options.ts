export interface OptionContract {
  type: "call" | "put";
  strikePrice: number;
  premium: number;
  quantity: number;
}

export interface AnalysisResult {
  priceAtExpiry: number;
  profitLoss: number;
}

export interface CombinedAnalysisResult {
  priceAtExpiry: number;
  [key: string]: number; // Index signature for dynamic keys
}

export interface OptionDetails {
  type: "call" | "put";
  strikePrice: number;
  longShort: "long" | "short";
  maxProfit: number | string;
  maxLoss: number | string;
  breakeven: number;
}
