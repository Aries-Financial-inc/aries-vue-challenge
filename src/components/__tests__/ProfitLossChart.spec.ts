import { shallowMount } from '@vue/test-utils'
import ProfitLossChart from '@/components/ProfitLossChart.vue'
import { sampleData } from '@/mocks/options'

describe('ProfitLossChart', () => {
  it('renders correctly', () => {
    const wrapper = shallowMount(ProfitLossChart, {
      propsData: {
        optionsData: sampleData
      }
    })
    expect(wrapper).toMatchSnapshot()
  })
})
