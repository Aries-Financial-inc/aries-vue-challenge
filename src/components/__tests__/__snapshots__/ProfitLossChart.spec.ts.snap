// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfitLossChart renders correctly 1`] = `
<div class="Root">
  <linechartjs-stub chartdata="[object Object]" chartoptions="[object Object]" datasetidkey="label" chartid="line-chart" width="400" height="400" cssclasses="" plugins=""></linechartjs-stub>
</div>
`;
